runtime/lua/vim/ui/img.lua

```diff
@@ -0,0 +1,557 @@
+    bytes = assert(vim.uv.fs_read(fd, 8, nil)) or ''
+    vim.uv.fs_close(fd, function() end)
+  end
+
+  local is_png = string.sub(bytes, 1, #PNG_SIGNATURE) == PNG_SIGNATURE
+  if is_png then
+    self.__metadata.format = 'PNG'
+  end
+  return is_png
+end
+
+---Converts this image into a PNG version.
+---Default compression is 7 and filter type is 5.
+---@param opts? {compression?:0|1|2|3|4|5|6|7|8|9, filter?:0|1|2|3|4|5|6|7|8|9, force?:boolean}
+---@return vim.ui.img.utils.Promise<vim.ui.Image>
+function M:into_png(opts)
+  opts = opts or {}
+  local promise = require('vim.ui.img.utils.promise').new({
+    context = 'image.into_png',
+  })
+
+  -- If true and trying to use cache, will force recreation
+  local force = opts.force or false
+  local compression = opts.compression
+  local filter_type = opts.filter
+
+  ---@type integer|nil
+  local quality = nil
+  if compression or filter_type then
+    quality = ((compression or 7) * 10) + ((filter_type or 5) % 10)
+  end
+
+  -- If image is a png, we are good to show it as is,
+  -- otherwise we'll need to convert it to a png
+  if self:is_png() and not force then
+    return promise:ok(self)
+  end
+
+  local cache_dir = M.cache_dir
+  assert(vim.fn.mkdir(cache_dir, 'p') == 1, 'failed to create ' .. cache_dir)
+
+  vim.uv.fs_realpath(
+    self.filename,
+    vim.schedule_wrap(function(err_realpath, path)
+      if err_realpath then
+        promise:fail(err_realpath)
+        return
+      end
+
+      -- Cached output image uses a sha256 of the full path to the original file
+      -- so we don't accidentally use a previously-cached image with a similar name
+      local out = vim.fs.joinpath(cache_dir, string.format('%s.png', vim.fn.sha256(path)))
+      local png = vim.ui.img.new(out)
+
+      -- Check if the image already exists
+      vim.uv.fs_stat(out, function(_, stat)
+        if stat and not force then
+          png
+            :into_png(opts)
+            :on_ok(function()
+              promise:ok(png)
+            end)
+            :on_fail(function(err)
+              promise:fail(err)
+            end)
+        else
+          self
+            :convert({ format = 'png', out = out, quality = quality })
+            :on_ok(function()
+              promise:ok(png)
+            end)
+            :on_fail(function(err)
+              promise:fail(err)
+            end)
+        end
+      end)
+    end)
+  )
+
+  return promise
+end
+
+---@class vim.ui.img.IdentifyOpts
+---@field format? boolean
+---@field size? boolean
+---@field [string] string|nil
+
+---@class vim.ui.img.Metadata
+---@field format? string
+---@field size? vim.ui.img.utils.Size
+---@field [string] string|nil
+
+---Returns information about the image by invoking ImageMagick's identify.
+---This will be some uppercase string like 'PNG'.
+---@param opts? vim.ui.img.IdentifyOpts
+---@return vim.ui.img.utils.Promise<vim.ui.img.Metadata>
+function M:identify(opts)
```

Comment:
this is definitely out of scope for this PR

Reply:
Only reason it existed was to easily find any image format's dimensions (width & height). If we just stick with PNG, I know how to read that from the file header. So, no problem.